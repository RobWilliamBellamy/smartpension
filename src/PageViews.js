import { Icon, Segment, Grid, Divider } from 'semantic-ui-react';
import { connect } from 'react-redux';

import PageViewsTable from './PageViewsTable';
import PageViewsChart from './PageViewsChart';
import PageViewsStatistics from './PageViewsStatistics';

import './css/WebPageViews.css';

/**
 * PageViews.
 * @returns view
 */
const PageViews = (props) => {
    
    const page = props.pages[props.index];

    return (<div>  
                <h1><Icon name='file outline' />{ page.tab.content }</h1>
                <p>{ page.tab.sub_title }</p>
                 
                <PageViewsTable name={ page.tab.id }
                                cols={[ { text: 'Page' },
                                        { text: 'Views', 
                                          desc: page.sort_desc, 
                                          sort: () => props.dispatch({ type: 'SORT_BY_VIEWS', data: props.index }) }]}  
                                data={[ Object.keys(page.key_values),
                                        Object.values(page.key_values) ]} 
                                footer={[ 'Total:', page.views_total ]} />

                <Segment>
                        <Grid columns={3}>
                                <Grid.Row>                                            
                                        <Grid.Column width={4}>
                                                <PageViewsChart data={ page.key_values } />
                                        </Grid.Column>  
                                        <Grid.Column width={2}></Grid.Column>                                      
                                        <Grid.Column width={10} verticalAlign='middle'>
                                                <PageViewsStatistics data={ page.stats } />                                                
                                        </Grid.Column>
                                </Grid.Row>
                        </Grid> 
                </Segment>               
            </div>);
};

/**
 * mapStateToProps
 * @param {*} state 
 * @returns 
 */
const mapStateToProps = state => ({
        pages: state
});

export default connect(mapStateToProps)(PageViews);