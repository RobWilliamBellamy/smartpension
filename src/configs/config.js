export const config = {
    logo: 'logo.png',
    default_page: '/pageviews',
    file_loader: {
        default: 'webserver.log',
        image: '/logo.png',
        accept: [".log"],
        placeHolder: 'Please select a web log file...',
        title: 'Upload Web Log',
        content: 'Please select a web log file to upload or use the default file from this project.'
    },
    errors: {
        file_parse_error: 'Failed to parse web log file'
    }
};