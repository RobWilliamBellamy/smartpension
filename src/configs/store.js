export const initial_state = [
    {
        tab: {
            id: 'page_views',
            path: '/pageviews',
            content: 'Page Views',
            sub_title: `Displays page views, change sort order by 
                clicking on the column header`
        },
        stats: {
            page_views: { label: 'Total Page Views', icon: 'eye', val: 0 },
            total_pages: { label: 'Total Pages', icon: 'sitemap', val: 0 }
        },
        sort_desc: true,            
        key_values: [],
        views_total: 0
    },
    {
        tab: {
            id: 'unique_views',
            path: '/uniqueviews',
            content: 'Unique Views',
            sub_title: `Displays unique page views, change sort order by 
                clicking on the column header`            
        },
        stats: {
            unique_views: { label: 'Total Unique Views', icon: 'eye', val: 0 },
            unique_users: { label: 'Total Unique Users', icon: 'user', val: 0 }
        },
        sort_desc: true,             
        key_values: [],
        views_total: 0
    }
];