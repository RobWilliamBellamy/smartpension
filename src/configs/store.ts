import { PageType } from './../types/PageType';

const store:Array<PageType> = [
    {
        tab: {
            id: 'page_views',
            path: '/pageviews',
            content: 'Page Views',
            sub_title: `Displays page views, change sort order by 
                clicking on the column header`
        },
        stats: [
            { label: 'Total Page Views', icon: 'eye', val: 0 },
            { label: 'Total Pages', icon: 'sitemap', val: 0 }
        ],
        sort_desc: true,            
        key_values: {},
        views_total: 0
    },
    {
        tab: {
            id: 'unique_views',
            path: '/uniqueviews',
            content: 'Unique Views',
            sub_title: `Displays unique page views, change sort order by 
                clicking on the column header`            
        },
        stats: [
            { label: 'Total Unique Views', icon: 'eye', val: 0 },
            { label: 'Total Unique Users', icon: 'user', val: 0 }
        ],
        sort_desc: true,             
        key_values: {},
        views_total: 0
    }
];

export default store;