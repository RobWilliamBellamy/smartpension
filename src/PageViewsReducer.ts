import { PageType, KeyValues } from './types/PageType';
import { PageViewsReducerActionType } from './types/PageViewsReducerActionType';

import initial_state  from './configs/store';
import { config } from './configs/config';

// PageViewsReducer, react to actions and modify state.
const pageViewsReducer = (state:Array<PageType> = initial_state , action:PageViewsReducerActionType) => {

    let new_state = Object.assign({}, state);
    switch (action.type) {
        case 'PROCESS_WEB_SERVER_LOG':             
            return parseLog(new_state, action.data);
        case 'SORT_BY_VIEWS':            
            return sortByViewCount(new_state, action.data);  
        default:
            return state;
    }
};

// Process the web server log.
const parseLog = (new_state:Array<PageType>, data:string) => {

    // Temp dictionary type.
    interface IDictionary {
        [key:string]: {
            views:number,
            unique:Array<string>
        }
    };

    const temp_data:IDictionary = {};
    const temp_users:Array<string> = [];

    try {
        // Parse the log file to seperate rows and columns.    
        const record_rows = data.split('\n');
        record_rows.forEach((rr, i) => {
        
            // Record web log as JSON.
            if (rr && rr.length > 0) {            
                
                const record_set = rr.split(' ');
                const page = record_set[0] as string;
                const ip_address = record_set[1] as string;
    
                // Record overall page views and unique views.
                let existing_page = temp_data[page];  
                temp_data[page] = (existing_page) ?                 
                { 
                    views: ++existing_page.views, 
                    unique: updateUniqueCount(existing_page.unique, ip_address)
                } : 
                { 
                    views: 1, 
                    unique: [ip_address]
                }; 
    
                // Record overall unique users.
                if (temp_users.indexOf(ip_address) === -1) {
                    temp_users.push(ip_address);
                }
            }
        }); 
    }
    catch (err) {
        throw new Error(config.errors.file_parse_error);
    }      

    // Short-hand references.
    const page_views = new_state[0];
    const unique_views = new_state[1];

    // Populate page view and unique view data.
    Object.keys(temp_data).forEach((d) => {       
        page_views.key_values[d] = temp_data[d].views;
        unique_views.key_values[d] = temp_data[d].unique.length;        
    });
    
    // Perform initial data sort.
    page_views.key_values = sortData(page_views.key_values, page_views.sort_desc);
    unique_views.key_values = sortData(unique_views.key_values, page_views.sort_desc);
    
    // Calculate view totals.
    page_views.views_total = calculateTotalViews(page_views.key_values);    
    unique_views.views_total = calculateTotalViews(unique_views.key_values);

    // Add stats.
    page_views.stats[0].val = page_views.views_total;
    page_views.stats[1].val = Object.keys(temp_data).length;                          
    unique_views.stats[0].val = unique_views.views_total;
    unique_views.stats[1].val = temp_users.length;

    return new_state;
};

// Update unique page view count.
const updateUniqueCount = (unique:Array<string>, ip_address:string):Array<string> => {
    
    if (unique.indexOf(ip_address) === -1) unique.push(ip_address);    
    return unique;
};

// Sort view data.
const sortData = (data:KeyValues, desc:boolean):KeyValues => {
    
    let tuples:Array<{key:string, val:string|number}> = [];
    for (const key in data) tuples.push({key, val: data[key]});

    tuples.sort((a, b) => {
        let av = a.val;
        let bv = b.val;
        return av < bv ? -1 : (av > bv ? 1 : 0);
    });

    // Reverse order if necessary.
    if (desc) {
        tuples.reverse();
    }

    let sorted_data:KeyValues = {};
    for (var i = 0; i < tuples.length; i++) {    
        sorted_data[tuples[i].key] = tuples[i].val;
    }
        
    return sorted_data;
};

// Calculate the views total.
const calculateTotalViews = (data:KeyValues):number => {

    const vals:Array<number> = Object.values(data)
                                     .map((i => parseInt(i as string)));    
    return vals.reduce((total:number, num:number) => {
        return total + num;
    }, 0);    
};

// Update sort by view count.
const sortByViewCount = (new_state:Array<PageType>, index:number) => {
    
    let page = new_state[index];
    let sort_desc = !page.sort_desc;

    page.sort_desc = sort_desc;
    page.key_values = sortData(page.key_values, sort_desc);
    
    return new_state;
};

export {
    pageViewsReducer, 
    parseLog,
    sortData
};