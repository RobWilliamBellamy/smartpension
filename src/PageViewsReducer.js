import { initial_state } from './configs/store.js';
import { config } from './configs/config.js';

// PageViewsReducer, react to actions and modify state.
const pageViewsReducer = (state = initial_state , action) => {

    let new_state = { ...state };
    switch (action.type) {
        case 'PROCESS_WEB_SERVER_LOG':             
            return processWebServerLog(new_state, action.data);
        case 'SORT_BY_VIEWS':            
            return sortByViewCount(new_state, action.data);  
        default:
            return state;
    }
};

// Process the web server log.
const processWebServerLog = (new_state, data) => {
        
    // Parse the log file to seperate rows and columns.
    const record_rows = data.split('\n');
    const temp_data = [];
    const temp_users = [];

    record_rows.forEach((rr, i) => {
        
        // Record web log as JSON.
        if (rr && rr.length > 0) {
            
            const record_set = rr.split(' ');
            const page = record_set[0];
            const ip_address = record_set[1];

            // Record overall page views and unique views.
            let existing_page = temp_data[page];  
            temp_data[page] = (existing_page) ?                 
            { 
                views: ++existing_page.views, 
                unique: updateUniqueCount(existing_page.unique, ip_address)
            } : 
            { 
                views: 1, 
                unique: [ip_address]
            }; 

            // Record overall unique users.
            if (temp_users.indexOf(ip_address) === -1) {
                temp_users.push(ip_address);
            }
        }
    });   

    // Short-hand references.
    const page_views = new_state[0];
    const unique_views = new_state[1];

    // Populate page view and unique view data.
    Object.keys(temp_data).forEach((d) => {
        page_views.key_values[d] = temp_data[d].views;
        unique_views.key_values[d] = temp_data[d].unique.length;        
    });
    
    // Perform initial data sort.
    page_views.key_values = sortViewData(page_views.key_values, page_views.sort_desc);
    unique_views.key_values = sortViewData(unique_views.key_values, page_views.sort_desc);
    
    // Calculate view totals.
    page_views.views_total = calculateTotalViews(page_views.key_values);    
    unique_views.views_total = calculateTotalViews(unique_views.key_values);

    // Add stats.
    page_views.stats.page_views.val = page_views.views_total;
    page_views.stats.total_pages.val = Object.keys(temp_data).length;                          
    unique_views.stats.unique_views.val = unique_views.views_total;
    unique_views.stats.unique_users.val = temp_users.length;

    return new_state;
};

// Update unique page view count.
const updateUniqueCount = (unique, ip_address) => {
    
    if (unique.indexOf(ip_address) === -1) unique.push(ip_address);    
    return unique;
};

// Sort view data.
const sortViewData = (data, desc) => {

    let tuples = [];
    for (var key in data) tuples.push([key, data[key]]);

    tuples.sort((a, b) => {
        a = a[1];
        b = b[1];
        return a < b ? -1 : (a > b ? 1 : 0);
    });

    // Reverse order if necessary.
    if (desc) {
        tuples.reverse();
    }

    let sorted_data = [];
    for (var i = 0; i < tuples.length; i++) {    
        sorted_data[tuples[i][0]] = tuples[i][1];
    }

    return sorted_data;
};

// Calculate the views total.
const calculateTotalViews = (data) => {

    return Object.values(data).reduce((total, num) => {
        return total + num;
    }, 0);
};

// Update sort by view count.
const sortByViewCount = (new_state, index) => {
    
    let page = new_state[index];
    let sort_desc = !page.sort_desc;

    page.sort_desc = sort_desc;
    page.key_values = sortViewData(page.key_values, sort_desc);
    
    return new_state;
};

export {
    pageViewsReducer, 
    processWebServerLog,  
};