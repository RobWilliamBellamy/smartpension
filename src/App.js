import React, { useEffect } from 'react';
import { BrowserRouter, Switch, Route, NavLink } from 'react-router-dom';
import { Tab, Segment, Header, Image } from 'semantic-ui-react';

import { createStore } from 'redux';
import { Provider } from 'react-redux';

import { loadFile } from './FileLoader';
import { pageViewsReducer }  from './PageViewsReducer';
import PageViews from './PageViews';

import { config } from './configs/config';

import './css/App.css';
import 'semantic-ui-css/semantic.min.css';

/**
 * defineTabs
 * @param {*} state 
 * @returns tabs
 */
 const defineTabs = (state) => {
    
    let panes = [];    
    state.forEach((page, i) => {
        
        const tab = page.tab;
        panes.push({
            menuItem: {
                as: NavLink,
                id: tab.id,
                content: tab.content,
                to: tab.path,
                exact: true,
                key: page.id
            },
            pane: (
                <Route path={ tab.path }
                       key={ tab.id }
                       exact                   
                       render={() => (
                            <Tab.Pane>
                                <PageViews index={ i }/>
                            </Tab.Pane>
                       )}
                />)
        });
    });

    return panes;
};

/**
 * App.
 * @returns app
 */
const App =() => {
        
    // Create Redux store.
    const store = createStore(pageViewsReducer);

    // Load the web server log file from disk.
    useEffect(() => {

        loadFile(config.web_server_log_file_path)
        .then((file) => {
            store.dispatch({ type: 'PROCESS_WEB_SERVER_LOG', data: file });
        })
        .catch(err => {
            console.log('Error loading data', err);
        });

    }, []);

    return (
        <Segment>
            <Segment>
                <Header>
                    <Image src='logo.png' size='large' floated='left'/>
                    Smart Pension                   
                    <Header.Subheader><b>Title:</b> FE Engineer Test</Header.Subheader>
                    <Header.Subheader><b>Author:</b> R.Bellamy</Header.Subheader>
                </Header>
            </Segment>
            <Provider store={ store }>
                <BrowserRouter>
                    <Switch>
                        <Tab key='menu' menu={ { tabular: true, className: 'topmenu' } }                             
                             renderActiveOnly={ false } panes={ defineTabs(store.getState()) } />                                  
                    </Switch>            
                </BrowserRouter>
            </Provider>
        </Segment>
    );
};

export default App;