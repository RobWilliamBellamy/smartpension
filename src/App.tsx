import { BrowserRouter, Switch } from 'react-router-dom';
import { Segment, Header, Image } from 'semantic-ui-react';

import { createStore } from 'redux';
import { Provider } from 'react-redux';

import AppContent from './AppContent';
import { pageViewsReducer }  from './PageViewsReducer';

import { config } from './configs/config';

import './css/App.css';
import 'semantic-ui-css/semantic.min.css';

/**
 * App
 * @returns app
 */
const App = () => {
    
    // Create Redux store.
    const store = createStore(pageViewsReducer);    
    
    return (<Segment>            
                <Segment>
                    <Header>
                        <Image src={ config.logo } size='large' floated='left'/>
                        Smart Pension                   
                        <Header.Subheader><b>Title:</b> FE Engineer Test</Header.Subheader>
                        <Header.Subheader><b>Author:</b> R.Bellamy</Header.Subheader>
                    </Header>
                </Segment>
                <Provider store={ store }>  
                    <BrowserRouter>
                        <Switch>  
                            <AppContent />
                        </Switch>
                    </BrowserRouter> 
                </Provider>
            </Segment>);
}

export default App;