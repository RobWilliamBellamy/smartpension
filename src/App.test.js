import { configure, shallow, mount } from "enzyme";
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";

import App from './App';
import { parseLog, sortData } from './PageViewsReducer';
import PageViews from './PageViews';

import initial_state from './configs/store';
import { config } from './configs/config';

const mockStore = configureStore([]);
configure({ adapter: new Adapter() });

let store;

// Create mock store for testing connected components.
beforeEach(() => {
    store = mockStore( initial_state );
});

describe('Test React App', () => {

    it("Renders without crashing", () => {
        shallow(<App />);
    });

    it("Renders the correct number of tabs", () => {
         const wrapper = mount(<App />);
         expect(wrapper.find(".topmenu").length)
         .toEqual(initial_state.length);
    });

    it('Renders PageViews correctly', () => {  
        const pageViews = renderer.create(
            <Provider store={ store }>
                <PageViews index={0}/>
            </Provider>);
        expect(pageViews.toJSON()).toMatchSnapshot();
    });
});

describe('Test reducer methods', () => {

    it("Parses web log", () => {
        expect(Array.isArray(parseLog(initial_state, 'A B\nC D')))
        .toBe(true);
    });
       
    it("Fails to parse web log", () => {
        expect(() => parseLog(initial_state, {}))
        .toThrow(config.errors.file_parse_error);
    });

    it("Sorts data", () => {
        let data = [];
        data['A'] = 0;
        data['B'] = 1;        
        expect(Object.keys(sortData(data, true)))
        .toEqual(['B', 'A']);
    });
});