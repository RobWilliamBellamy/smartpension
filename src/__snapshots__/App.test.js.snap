// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test React App Renders PageViews correctly 1`] = `
<div>
  <h1>
    <i
      aria-hidden="true"
      className="file outline icon"
      onClick={[Function]}
    />
    Page Views
  </h1>
  <p>
    Displays page views, change sort order by 
                clicking on the column header
  </p>
  <table
    className="ui celled sortable striped table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
        >
          Page
        </th>
        <th
          className="descending sorted"
          onClick={[Function]}
        >
          Views
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    />
    <tfoot
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
        >
          <b>
            Total:
          </b>
        </th>
        <th
          className=""
        >
          <b>
            0
          </b>
        </th>
      </tr>
    </tfoot>
  </table>
  <div
    className="ui segment"
  >
    <div
      className="ui three column grid"
    >
      <div
        className="row"
      >
        <div
          className="four wide column"
        >
          <div
            className="VictoryContainer"
            style={
              Object {
                "height": "100%",
                "pointerEvents": "none",
                "position": "relative",
                "touchAction": "none",
                "width": "100%",
              }
            }
          >
            <svg
              height={350}
              role="img"
              style={
                Object {
                  "height": "100%",
                  "pointerEvents": "all",
                  "width": "100%",
                }
              }
              viewBox="0 0 350 350"
              width={350}
            >
              <g />
            </svg>
            <div
              style={
                Object {
                  "height": "100%",
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": "100%",
                  "zIndex": 99,
                }
              }
            >
              <svg
                height={350}
                style={
                  Object {
                    "height": "100%",
                    "overflow": "visible",
                    "width": "100%",
                  }
                }
                viewBox="0 0 350 350"
                width={350}
              />
            </div>
          </div>
        </div>
        <div
          className="two wide column"
        />
        <div
          className="middle aligned ten wide column"
        >
          <div
            className="ui small two statistics"
            name="statistics_page_views"
          >
            <div
              className="ui left floated horizontal statistic"
            >
              <div
                className="value"
              >
                <i
                  aria-hidden="true"
                  className="eye icon"
                  onClick={[Function]}
                />
                 
                0
              </div>
              <div
                className="label"
              >
                Total Page Views
              </div>
            </div>
            <div
              className="ui left floated horizontal statistic"
            >
              <div
                className="value"
              >
                <i
                  aria-hidden="true"
                  className="sitemap icon"
                  onClick={[Function]}
                />
                 
                0
              </div>
              <div
                className="label"
              >
                Total Pages
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
